### PR TITLE
Added backoff to fetcher worker to partially address #746

### PR DESCRIPTION
**Pull request by mhutchinson**
This is a huge improvement from retying without any backoff. However, this could still benefit from:
* being tunable
* supporting hard failing if requests can't be completed within the backoff configuration
* resetting the backoff if requests start succeeding faster

### CODE REVIEW

File: scanner/fetcher.go [hunk 1]
```diff
@@ -257,15 +257,26 @@ func (f *Fetcher) runWorker(ctx context.Context, ranges <-chan fetchRange, fn fu
...
 		for r.start <= r.end {
-			// Fetcher.Run() can be cancelled while we are looping over this job.
-			if err := ctx.Err(); err != nil {
-				glog.Warningf("%s: Worker context closed: %v", f.uri, err)
-				return
+			// TODO(pavelkalinnikov): Make these parameters tunable.
+			// This backoff will only apply to a single request and be reset for the next one.
+			// This precludes reaching some kind of stability in request rate, but means that
+			// an intermittent problem won't harm long-term running of the worker.
+			workerBackoff := &backoff.Backoff{
```
Comment (pavelkalinnikov) [inline scanner/fetcher.go:264]: nit: No need in "worker" in this name, since it spans a single request. How about just "b" or "bo"?

**Comment by mhutchinson** [comment]
Latest commit addresses "resetting the backoff if requests start succeeding faster" from the initial description. This is still incrementally much better (and with no downsides compared to the original?), but not ideal.

File: scanner/fetcher.go [hunk 2]
```diff
@@ -257,15 +257,26 @@ func (f *Fetcher) runWorker(ctx context.Context, ranges <-chan fetchRange, fn fu
...
+
+			var resp *ct.GetEntriesResponse
+			if err := workerBackoff.Retry(ctx, func() error {
+				var err error
+				resp, err = f.client.GetRawEntries(ctx, r.start, r.end)
+				return err
+			}); err != nil {
 				glog.Errorf("%s: GetRawEntries() failed: %v", f.uri, err)
-				// TODO(pavelkalinnikov): Introduce backoff policy and pause here.
+				// TODO(mhutchinson): This needs to be able to signal a failure.
```
Comment (pavelkalinnikov) [inline scanner/fetcher.go:278]: I think the best way to signal the failure is to decorate `LogClient`. So, the client will know about the error before they get to any retries, or this line here. They can also know errors caused by each retry.

How about moving this TODO to line 272, saying something like: "Retry requests and report errors in a LogClient decorator instead."?

**Review by pavelkalinnikov — APPROVED** [review]

File: scanner/fetcher.go (hunk 2 above)
Comment (mhutchinson) [reply scanner/fetcher.go]: I've moved the TODO to there. The current approach maintains the old approach of continually looping forever. I'm heavily optimizing for keeping this PR as localized as possible, so not adding an error channel yet.

**Pull request merged**